Snapshots can be captured as store-backed Directories as well as tar files.

illicitonion opened the pull request:
A follow-up PR will remove the tar file implementation, and with it, a lot of code.

Moves Directory traversal logic from fs_util into fs.

Introduces a GetFileDigest trait which can be implemented both by Context (for memoized graph operations) and in fs_util and tests (for one-off file reading).

illicitonion commented:
Depends on https://github.com/pantsbuild/pants/pull/5104

illicitonion commented on src/rust/engine/fs/src/snapshot.rs:
```diff
@@ -0,0 +1,352 @@
+        }
+      } else {
+        dir_futures.push(
+          // TODO: Memoize this in the graph
```
This would be nice to do at some point, but I'm not going to rust to do it, because it's no worse than the current behaviour.

stuhood replied:
Yea, this can wait.

illicitonion commented on src/rust/engine/fs/src/snapshot.rs:
```diff
@@ -0,0 +1,352 @@
+  }
+
+  // Ripped from lib.rs's test module
+  // TODO: Work out how to share things across tests
```
Thoughts welcome

stuhood replied:
I think that that would require a brand new crate, imported only inside the test module?

illicitonion commented on src/rust/engine/fs/src/snapshot.rs:
```diff
@@ -0,0 +1,352 @@
+    file.set_permissions(permissions).unwrap();
+  }
+
+  // TODO: Is there a better way to do this?
```
I couldn't find an obvious "Down-cast to known enum type" feature, but it could be that I'm just not using the right terminology... Any pointers would be appreciated :)

stuhood replied:
I don't see a way either. https://stackoverflow.com/a/45473824

But it looks like in this case you don't really need to verify the type of the PathStat, since you just created the file/dir. So a bigger boilerplate-remover in here would be to add a `fn stat(p: &Path) -> Vec<PathStat>` test helper to replace both cast functions. I think you could do the equivalent of `VFS::expand(PathGlobs::create([p.to_string], [])).wait()`... maybe overkill, but not likely to slow the tests down a noticeable amount.

stuhood commented on src/rust/engine/fs/fs_util/src/main.rs:
```diff
@@ -273,120 +280,22 @@ fn make_posix_fs<P: AsRef<Path>>(root: P) -> fs::PosixFS {
+    let file_copy = file.clone();
+    let store = self.0.clone();
+    self
+      .1
```
This probably warrants naming the fields and adding a constructor... sorry!

stuhood commented on src/rust/engine/fs/fs_util/src/main.rs:
```diff
@@ -273,120 +280,22 @@ fn make_posix_fs<P: AsRef<Path>>(root: P) -> fs::PosixFS {
+    let store = self.0.clone();
+    self
+      .1
+      .clone()
```
Shouldn't need to clone here... `read_file` takes the `PosixFS` by reference.

stuhood commented on src/rust/engine/fs/src/snapshot.rs:
```diff
@@ -1,20 +1,352 @@
+            // Because there are no children of this Dir, it must be empty.
+            dir_futures.push(
+              futures::done(
+                store
```
The `store.record_directory` call should happen on a `CpuPool` in order to not block the main thread. Let's not do it in this PR, as I suspect that it will take some fidgeting to get a reference to the pool into `Store`. But a TODO+ticket would be good.

illicitonion commented:
Rebased and addressed all comments.

I'll rebase on top of #5117 when it lands, then land this. This PR doesn't actually start putting stuff in lmdb, it just adds the code to allow doing so.

When I eventually land https://github.com/pantsbuild/pants/pull/5106 it will do a clean cut-over and atomically stop writing to tar files and start writing to lmdb.

stuhood commented:
Rebased atop #5117.
